src/mesh/interpolation/hermite_spline_z.cxx

```diff
@@ -62,20 +62,23 @@ void ZHermiteSpline::calcWeights(const Field3D& delta_z, const std::string& regi
 
     // The integer part of zt_prime are the indices of the cell
     // containing the field line end-point
-    k_corner[i.ind] = static_cast<int>(floor(delta_z(x, y, z)));
+    int corner_zind = floor(delta_z(x, y, z));
 
     // t_z is the normalised coordinate \in [0,1) within the cell
     // calculated by taking the remainder of the floating point index
-    const BoutReal t_z = delta_z(x, y, z) - static_cast<BoutReal>(k_corner[i.ind]);
+    const BoutReal t_z = delta_z(x, y, z) - static_cast<BoutReal>(corner_zind);
+
+    // make corner_zind be in the range 0<=corner_zind<nz
+    corner_zind = ((corner_zind % ncz) + ncz) % ncz;
 
-    // make k_corner be in the range 0<=k_corner<nz
-    k_corner[i.ind] = ((k_corner[i.ind] % ncz) + ncz) % ncz;
+    // Convert z-index to Ind3D
+    k_corner[i.ind] = Ind3D((i.x()*ncy + i.y())*ncz + corner_zind, ncy, ncz);
```

Comment:
This is because we want the `(x, y)` indices from the current loop iteration but we _know_ the z index?

Reply:
Yes - basically we want `k_corner(x, y, z) = {x, y, floor(delta_z(x, y, z))}

I notice now I should have used `x` and `y` instead of `i.x()` and `i.y()`, since they're there already.